internal/store/store.go

```diff
@@ -13,34 +13,114 @@ import (
 // number and drop older data once that threshold is exceeded. All functions
 // are thread safe.
 type Store struct {
-	mu       sync.RWMutex
-	timeTree *avltree.Tree
-	size     int
+	mu           sync.RWMutex
+	size         int
+	maxPerSource int
+
+	sourceIDs map[string]*avltree.Tree
+
+	// oldestValueTree stores each tree's oldest value for pruning. As data is
+	// added and needs to be pruned, it is done so from here.
+	oldestValueTree *treeStorage
+
+	// count is incremented each Put. It is used to determine when to prune. When
+	// an envelope is pruned, it is decremented.
+	count int
 }
 
 // NewStore creates a new store.
-func NewStore(size int) *Store {
+func NewStore(size, maxPerSource int) *Store {
 	return &Store{
-		timeTree: avltree.NewWith(utils.Int64Comparator),
-		size:     size,
+		size:            size,
+		maxPerSource:    maxPerSource,
+		sourceIDs:       make(map[string]*avltree.Tree),
+		oldestValueTree: newTreeStorage(),
 	}
 }
 
 // Put adds a batch of envelopes into the store.
 func (s *Store) Put(envs []*loggregator_v2.Envelope) {
 	s.mu.Lock()
 	defer s.mu.Unlock()
-
 	for _, e := range envs {
-		s.timeTree.Put(e.Timestamp, e)
 
-		if s.timeTree.Size() > s.size {
-			// Truncate the oldest envelope
-			s.timeTree.Remove(s.timeTree.Left().Key)
+		t, ok := s.sourceIDs[e.SourceId]
```

Comment:
I think it would be good to have two separate locks here. One around the map and one around the `Put` to the tree. We can write some benchmarks and add that later though if really needed.